dap/debugger.go

```diff
@@ -0,0 +1,117 @@
+package dap
+
+import (
+	"path/filepath"
+	"sync"
+
+	"github.com/ysugimoto/falco/ast"
+	"github.com/ysugimoto/falco/interpreter"
+)
+
+type Debugger struct {
+	mode    interpreter.DebugState
+	stateCh <-chan interpreter.DebugState
+
+	printFunc         func(msg string)
+	notifyStoppedFunc func(params *notifyStoppedEventParams)
+
+	breakpoints *breakpointColl
+	stacks      *stackColl
+}
+
+func newDebugger(stateCh <-chan interpreter.DebugState) *Debugger {
+	return &Debugger{
+		stateCh: stateCh,
+		breakpoints: &breakpointColl{
+			breakpoints: map[string][]breakpoint{},
+			counter:     0,
+			mu:          sync.Mutex{},
+		},
+		stacks: &stackColl{
+			stacks:  []stack{},
+			counter: 0,
+			mu:      sync.Mutex{},
+		},
+	}
+}
+
+func (d *Debugger) Run(node ast.Node) interpreter.DebugState {
+	switch d.mode {
+	case interpreter.DebugStepIn, interpreter.DebugStepOver:
+		d.appendStack(node)
+		d.notifyStoppedFunc(&notifyStoppedEventParams{
+			reason: "step",
+		})
+		return d.waitForNewState()
+	case interpreter.DebugStepOut:
+		d.mode = interpreter.DebugStepOver
+		d.appendStack(node)
+		d.notifyStoppedFunc(&notifyStoppedEventParams{
+			reason: "step",
+		})
+
+		return d.waitForNewState()
+	default:
+		if bp := d.getBreakpoint(node); bp != nil {
+			d.mode = interpreter.DebugStepOver
+			d.appendStack(node)
+			d.notifyStoppedFunc(&notifyStoppedEventParams{
+				reason:        "breakpoint",
+				breakpointIDs: []int{bp.id},
+			})
+			return d.waitForNewState()
+		}
+		return interpreter.DebugPass
+	}
+}
+
+func (d *Debugger) Message(msg string) {
+	d.printFunc(msg)
+}
+
+func (d *Debugger) waitForNewState() interpreter.DebugState {
+	d.mode = <-d.stateCh
+
+	return d.mode
+}
+
+func (d *Debugger) getBreakpoint(node ast.Node) *breakpoint {
+	meta := node.GetMeta()
+
+	path, err := filepath.Abs(meta.Token.File)
+	if err != nil {
+		// TODO: handle error
```

Comment:
How should we treat for the error?
In terms of `ast.Node`, falco always treat as absolute path so `meta.Token.File` always has an absolute path.
Therefore you don't need to get the absolute path using `filepath.Abs`.

Additionally, falco may retrieves Fastly remote reources like `acl`, `backend`, etc then the remote source file will be `Remote.Acl.[name]` and the the getBreakpoint` works fine even not a filepath?

Reply:
Thank you for your comment. I've just fixed it!

> Additionally, falco may retrieves Fastly remote reources like `acl`, `backend`, etc then the remote source file will be `Remote.Acl.[name]` and the the getBreakpoint` works fine even not a filepath?

Currently, it doesn't matter because `dap` subcommand does not support remote resources feature.
We need to implement it as a future work.